config/routes.rb

```diff
@@ -1583,6 +1583,11 @@ def patch(*)
 
     get "/theme-qunit" => "qunit#theme"
 
+    # This is a special route that is used when theme QUnit tests are run through testem which appends a testem_id to the
+    # path. Unfortunately, testem's proxy support does not allow us to easily remove this from the path, so we have to
+    # handle it here.
+    get "/:testem_id/theme-qunit" => "qunit#theme" if Rails.env.development?
```

Comment:
I worry this could cause us headaches in future. If some route with lower precedence is defined like

```
get "/something/:slug"
```

and then a user sets the slug to `theme-qunit`, then that route would be inaccessible because `/:testem_id/theme-qunit` would match first.

How about we tell `testem` to proxy all `theme-qunit` requests to `/theme-qunit/...`. That way, the prefix is the only important thing. Something along the lines of

```diff
diff --git a/app/assets/javascripts/discourse/testem.js b/app/assets/javascripts/discourse/testem.js
index 34257f867f..8c656be089 100644
--- a/app/assets/javascripts/discourse/testem.js
+++ b/app/assets/javascripts/discourse/testem.js
@@ -161,6 +161,10 @@ const themeTestPages = process.env.THEME_TEST_PAGES;
 if (themeTestPages) {
   module.exports.test_page = themeTestPages.split(",");
   module.exports.proxies = {};
+  module.exports.proxies[`/*/theme-qunit`] = {
+    target: `${target}/theme-qunit`,
+    xfwd: true,
+  };
   module.exports.proxies["/*/*"] = { target, xfwd: true };
 
   module.exports.middleware = [
diff --git a/config/routes.rb b/config/routes.rb
index 6f5a7656b3..a563ef54dd 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -1582,11 +1582,7 @@ Discourse::Application.routes.draw do
     post "/safe-mode" => "safe_mode#enter", :as => "safe_mode_enter"
 
     get "/theme-qunit" => "qunit#theme"
-
-    # This is a special route that is used when theme QUnit tests are run through testem which appends a testem_id to the
-    # path. Unfortunately, testem's proxy support does not allow us to easily remove this from the path, so we have to
-    # handle it here.
-    get "/:testem_id/theme-qunit" => "qunit#theme" if Rails.env.development?
+    get "/theme-qunit/*_ignored_testem_path" => "qunit#theme"
 
     post "/push_notifications/subscribe" => "push_notification#subscribe"
     post "/push_notifications/unsubscribe" => "push_notification#unsubscribe"
```

Reply:
Good idea, I made the route more specific instead.

```
    # This is a special route that is used when theme QUnit tests are run through testem which appends a testem_id to the
    # path. Unfortunately, testem's proxy support does not allow us to easily remove this from the path, so we have to
    # handle it here.
    if Rails.env.test? || Rails.env.development?
      get "/testem-theme-qunit/:testem_id/theme-qunit" => "qunit#theme",
          :constraints => {
            testem_id: /\d+/,
          }
    end
```